lib/pure/options.nim

```diff
@@ -215,6 +215,21 @@ proc get*[T](self: Option[T], otherwise: T): T {.inline.} =
   else:
     otherwise
 
+proc get*[T](self: Option[T], val: var T): bool {.inline.} =
```

Comment:
But shouldn't that be `val: out T`?

Reply:
I'm sorry, but what difference does that do? Is it just a naming standard or does it actually do something? Sorry, I didn't even know that prefix existed before now.

Reply:
`out` vs `var` parameters: https://nim-lang.org/docs/manual_experimental.html#strict-definitions-and-nimout-parameters-nimout-parameters

Reply:
Should it? Wouldn't that require `val` to be set before returning? In this case you'd need a `val = default(T)` in the `isNone` case which doesn't really make much sense.

Reply:
As opposed to what though? Otherwise you need the same `val = default(T)` on the client side. Keep in mind that `--experimental:strictDefs` is coming.

Reply:
`val` is not always assigned so it can't be `out T`.

Reply:
Yes well, it has to be always assigned then.

Reply:
Should there be a `default` optional argument, in that case? What if the user doesn't provide that argument? Should we just raise a Defect?

Reply:
But your `unpackInto` models an `out` parameter moreso than it models a `var` parameter.

Reply:
The entire gist of this function is:
- you have an empty value, say an `int`
- you have an `Option[int]`, you don't know whether it contains something or not
- you can call this function, if the `Option[int]` wasn't empty, `true` will be returned, and you can throw that into an if statement
- else, you now know that nothing was done to the empty value and also know that the `Option[int]` was empty

an `out` parameter doesn't make sense here if we're returning a `bool`, which we'll use for a conditional check

Reply:
> But your `unpackInto` models an `out` parameter moreso than it models a `var` parameter.

Not really. The concept is that in changes the output to the result of the option and returns whether or not it was set. As ElegantBeef showed this can be used to provide an override function where available options will override the currently assigned value or otherwise leave it alone. This isn't just the current `get` without a return value, this is an unpacking of the someness of the option and the value.

You can also use it for a fun option walrus:

```nim
template `?=`[T](x: untyped, y: Option[T]): bool =
  var x: T
  unpack(y, x)
  
var x = some(100)

if y ?= x:
  echo y # This variable is only defined in the scope where the option is valid
```

Reply:
The sequence `var x: T; unpack(y, x)` does not compile under `--experimental:strictDefs` when `unpack` does not take an `out` parameter.

Reply:
In that case, `--experimental:strictDefs` is deeply flawed. You can't just make 90% of code invalid in the name of "strictness" or "safety". `x` won't even be accessible outside the scope, and since the boolean check returns false in the case of an empty option, there's not gonna be any real problems, unless you're doing some black magic.

I do not want to sound rude or undermining, but I'd really revise `strictDefs` before imposing it on everything given how much flexible syntax it renders completely impossible. That was really the magic of Nim that attracted me to learn it and taking it away is a huge slap in the face for a lot of people. This seems like a *great* idea, but the execution needs to be revised.

Reply:
strictDefs means that where you did `var x: T` you sometimes need to write `var x = default(T)` instead. Still seems pretty reasonable to me given that it found real bugs in my own real code and I'm not that bad at Nim programming.